File: src/main/java/com/fengsheng/handler/remove_robot_tos.java
package com.fengsheng.handler;

import com.fengsheng.Game;
import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.RobotPlayer;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class remove_robot_tos extends AbstractProtoHandler<Fengsheng.remove_robot_tos> {
    private static final Logger log = Logger.getLogger(remove_robot_tos.class);

    @Override
    protected void handle0(HumanPlayer player, Fengsheng.remove_robot_tos pb) {
        Player[] players = player.getGame().getPlayers();
        RobotPlayer robotPlayer = null;
        synchronized (Game.class) {
            for (Player p : player.getGame().getPlayers()) {
                if (p instanceof RobotPlayer) {
                    robotPlayer = (RobotPlayer) p;
                    players[p.location()] = null;
                    break;
                }
            }
        }
        if (robotPlayer != null) {
            log.info(robotPlayer.getPlayerName() + "离开了房间");
            var reply = Fengsheng.leave_room_toc.newBuilder().setPosition(robotPlayer.location()).build();
            for (Player p : players) {
                if (p instanceof HumanPlayer) {
                    ((HumanPlayer) p).send(reply);
                }
            }
        }
    }
}

File: src/main/java/handler/remove_robot_tos.kt
package com.fengsheng.handler

import com.fengsheng.Game
import com.fengsheng.HumanPlayer
import com.fengsheng.RobotPlayer
import com.fengsheng.protos.Fengsheng
import com.fengsheng.protos.Fengsheng.leave_room_toc

org.apache.log4j.Logger
class remove_robot_tos : AbstractProtoHandler<Fengsheng.remove_robot_tos?>() {
    override fun handle0(player: HumanPlayer, pb: Fengsheng.remove_robot_tos?) {
        val players = player.game.players
        var robotPlayer: RobotPlayer? = null
        synchronized(Game::class.java) {
            for (p in player.game.players) {
                if (p is RobotPlayer) {
                    robotPlayer = p
                    players[p.location()] = null
                    break
                }
            }
        }
        if (robotPlayer != null) {
            log.info(robotPlayer!!.playerName + "离开了房间")
            val reply = leave_room_toc.newBuilder().setPosition(robotPlayer!!.location()).build()
            for (p in players) {
                if (p is HumanPlayer) {
                    p.send(reply)
                }
            }
        }
    }

    companion object {
        private val log = Logger.getLogger(remove_robot_tos::class.java)
    }
}